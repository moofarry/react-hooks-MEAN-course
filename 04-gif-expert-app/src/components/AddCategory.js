import React from "react";
import PropTypes from "prop-types";

export const AddCategory = ({ setCategories }) => {
  const [inputValue, setInputValue] = React.useState('');

  const handleInputChange = (e) => {
    setInputValue(e.target.value);
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    if (inputValue.trim().length > 2) {
      setCategories((prevCategories) => [ inputValue,...prevCategories]);
      setInputValue("");
    }
  };

  return (
    <form onSubmit={handleSubmit}>
      <input
        value={inputValue}
        type="text"
        onChange={handleInputChange}
      ></input>
    </form>
  );
};

AddCategory.propTypes = {
  setCategories: PropTypes.func.isRequired,
};
