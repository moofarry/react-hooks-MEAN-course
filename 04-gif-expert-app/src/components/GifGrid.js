import React from "react";

export const GifGrid = ({ category }) => {
  const [count, setCount] = React.useState(0);

  React.useEffect(() => {
    getGifs();
  });

  const getGifs = async () => {
    const url =
      "https://api.giphy.com/v1/gifs/search?q=Uribe&limit=10&api_key=Enhu5gM26OHBLH0hYgf1tKNVcT2JHheo";
    const resp = await fetch(url);
    const { data } = await resp.json();

    const gifs = data.map(({ id, title, images }) => {
      return {
        id: id,
        title: title,
        url: images?.downsized_medium.url,
      };
    });

    console.log(`gifs`, gifs);
  };
  return (
    <div>
      <h3>{category}</h3>

      <h3>{count}</h3>
      <button
        onClick={() => {
          setCount(count + 1)}}
      ></button>
    </div>
  );
};
