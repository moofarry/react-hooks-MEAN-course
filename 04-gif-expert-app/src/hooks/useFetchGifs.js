import React from "react";
import { getGifs } from "../helpers/getGif";

export const useFetchGifs = (category) => {
  const [state, setstate] = React.useState({
    data: [],
    loading: true,
  });

  React.useEffect(() => {
    getGifs(category).then((imgs) => {
      setstate({
        data: imgs,
        loading: false,
      });
    });
  }, [category]);

  return state;
};
