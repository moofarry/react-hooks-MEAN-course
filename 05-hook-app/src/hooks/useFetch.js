import { useState, useEffect, useRef } from "react";

export const useFetch = (url) => {
  const isMounted = useRef(true);
  const [state, setState] = useState({
    loading: true,
    error: null,
    data: null,
  });

  useEffect(() => {
    isMounted.current = true;

    return () => {
      isMounted.current = false;
    };
  }, []);

  useEffect(() => {
    fetch(url)
      .then((resp) => resp.json())
      .then((data) => {
        if (isMounted.current) {
          setState({ loading: false, error: null, data });
        }
      })
      .catch(() => {
        setState({
          data: null,
          loading: false,
          error: "could not load the data",
        });
      });
  }, [url]);

  return state;
};
