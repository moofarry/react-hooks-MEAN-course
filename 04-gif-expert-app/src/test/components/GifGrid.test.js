import React from "react";
import "@testing-library/jest-dom";

import { shallow } from "enzyme";
import { GifGrid } from "../../components/GifGrid";
import { useFetchGifs } from "../../hooks/useFetchGifs";

jest.mock("../../hooks/useFetchGifs");

describe("test in GifGrid", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  const category = "Naruto";

  test("should show correctly", () => {
    useFetchGifs.mockReturnValue({
      data: [],
      loading: true,
    });

    const wrapper = shallow(<GifGrid category={category} />);
    expect(wrapper).toMatchSnapshot();
  });

  test("should show items when img is load useFetchGifs", () => {
    const gifs = [
      {
        id: "ABC",
        url: "https://google.com",
        title: "ABC Title",
      },
    ];

    useFetchGifs.mockReturnValue({
      data: gifs,
      loading: false,
    });

    const wrapper = shallow(<GifGrid category={category} />);
    expect(wrapper).toMatchSnapshot();
  });
});
