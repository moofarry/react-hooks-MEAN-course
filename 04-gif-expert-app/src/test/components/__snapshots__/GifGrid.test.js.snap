// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test in GifGrid should show correctly 1`] = `
<Fragment>
  <h3>
    Naruto
  </h3>
  <p
    className="animate__animated animate__heartBeat"
  >
    Loading
  </p>
  <div
    className="card-grid animate__animated animate__bounce"
  />
</Fragment>
`;

exports[`test in GifGrid should show items when img is load useFetchGifs 1`] = `
<Fragment>
  <h3>
    Naruto
  </h3>
  <div
    className="card-grid animate__animated animate__bounce"
  >
    <GifGridItem
      id="ABC"
      key="ABC"
      title="ABC Title"
      url="https://google.com"
    />
    <GifGridItem
      id="ABCD"
      key="ABCD"
      title="ABC Title"
      url="https://google.com"
    />
  </div>
</Fragment>
`;
